File: clients/fsharp/Model/DebugState.fs
#nowarn "0058"

namespace Komus24
namespace Komus24.Model

open Komus24

/// TODO - Document
type DebugState = {
    /// TODO - Document
    PressedKeys: string[];
} with

    /// Write DebugState to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.PressedKeys.Length
        this.PressedKeys |> Array.iter (fun value ->
            let valueData : byte[] = System.Text.Encoding.UTF8.GetBytes value
            writer.Write valueData.Length
            writer.Write valueData )
        ()

    /// Read DebugState from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        PressedKeys = [|for _ in 1 .. reader.ReadInt32() do
                          yield reader.ReadInt32() |> reader.ReadBytes |> System.Text.Encoding.UTF8.GetString |]
    }

File: clients/csharp/Model/DebugState.cs
namespace Komus24.Model
{
    /// <summary>
    /// TODO - Document
    /// </summary>
    public struct DebugState
    {
        /// <summary>
        /// TODO - Document
        /// </summary>
        public string[] PressedKeys { get; set; }
    
        public DebugState(string[] pressedKeys)
        {
            this.PressedKeys = pressedKeys;
        }
    
        /// <summary> Read DebugState from reader </summary>
        public static DebugState ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new DebugState();
            result.PressedKeys = new string[reader.ReadInt32()];
            for (int pressedKeysIndex = 0; pressedKeysIndex < result.PressedKeys.Length; pressedKeysIndex++)
            {
                result.PressedKeys[pressedKeysIndex] = System.Text.Encoding.UTF8.GetString(reader.ReadBytes(reader.ReadInt32()));
            }
            return result;
        }
    
        /// <summary> Write DebugState to writer </summary>
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(PressedKeys.Length);
            foreach (var pressedKeysElement in PressedKeys)
            {
                var pressedKeysElementData = System.Text.Encoding.UTF8.GetBytes(pressedKeysElement);
                writer.Write(pressedKeysElementData.Length);
                writer.Write(pressedKeysElementData);
            }
        }
    
        /// <summary> Get string representation of DebugState </summary>
        public override string ToString() {
            string stringResult = "DebugState { ";
            stringResult += "PressedKeys: ";
            stringResult += "[ ";
            int pressedKeysIndex = 0;
            foreach (var pressedKeysElement in PressedKeys)
            {
                if (pressedKeysIndex != 0) {
                    stringResult += ", ";
                }
                stringResult += "\"" + pressedKeysElement + "\"";
                pressedKeysIndex++;
            }
            stringResult += " ]";
            stringResult += " }";
            return stringResult;
        }
    }
}